In [1]:
# Below is the endpoint for the Rat inspection endpoint
# 'https://data.cityofnewyork.us/resource/p937-wjvj.json'

# Below is the endpoint for the dog dataset
# 'https://data.cityofnewyork.us/resource/nu7n-tubp.json'

In [2]:
import requests
import pandas as pd
import numpy as np
import re
import altair as alt
import statsmodels.formula.api as smf
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
#import Stargazer

In [3]:
# r = requests.get('https://data.cityofnewyork.us/resource/p937-wjvj.json')
# r.status_code

In [4]:
# rat_df = pd.DataFrame(r.json())
# rat_df.head()

In [5]:
new_york_listings = pd.read_csv('http://data.insideairbnb.com/united-states/ny/new-york-city/2021-04-07/data/listings.csv.gz', compression='gzip')
new_york_listings['City'] = 'New York, New York'
new_york_listings.head()

id                        listing_url       scrape_id last_scraped  \
0  2595  https://www.airbnb.com/rooms/2595  20210407173759   2021-04-09   
1  3831  https://www.airbnb.com/rooms/3831  20210407173759   2021-04-12   
2  5121  https://www.airbnb.com/rooms/5121  20210407173759   2021-04-09   
3  5136  https://www.airbnb.com/rooms/5136  20210407173759   2021-04-10   
4  5178  https://www.airbnb.com/rooms/5178  20210407173759   2021-04-12   

                                                name  \
0                              Skylit Midtown Castle   
1  Whole flr w/private bdrm, bath & kitchen(pls r...   
2                                    BlissArtsSpace!   
3           Spacious Brooklyn Duplex, Patio + Garden   
4                   Large Furnished Room Near B'way　   

                                         description  \
0  Beautiful, spacious skylit studio in the heart...   
1  Enjoy 500 s.f. top floor in 1899 brownstone, w...   
2  <b>The space</b><br />HELLO EVERYONE AND THANK...   
3  We welcome you to stay in our lovely 2 br dupl...   
4  Please don’t expect the luxury here just a bas...   

                               neighborhood_overview  \
0  Centrally located in the heart of Manhattan ju...   
1  Just the right mix of urban center and local n...   
2                                                NaN   
3                                                NaN   
4    Theater district, many restaurants around here.   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/f0813a11-40b2...     2845   
1  https://a0.muscache.com/pictures/e49999c2-9fd5...     4869   
2  https://a0.muscache.com/pictures/2090980c-b68e...     7356   
3  https://a0.muscache.com/pictures/a3f7ace8-e7a9...     7378   
4  https://a0.muscache.com/pictures/12065/f070997...     8967   

                                 host_url  ... review_scores_location  \
0  https://www.airbnb.com/users/show/2845  ...                   10.0   
1  https://www.airbnb.com/users/show/4869  ...                   10.0   
2  https://www.airbnb.com/users/show/7356  ...                    9.0   
3  https://www.airbnb.com/users/show/7378  ...                    8.0   
4  https://www.airbnb.com/users/show/8967  ...                   10.0   

  review_scores_value license instant_bookable calculated_host_listings_count  \
0                 9.0     NaN                f                              3   
1                10.0     NaN                f                              1   
2                 9.0     NaN                f                              1   
3                10.0     NaN                f                              1   
4                 9.0     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           3   
1                                           1   
2                                           0   
3                                           1   
4                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            1   
3                                            0   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  \
0                                           0              0.35   
1                                           0              4.98   
2                                           0              0.35   
3                                           0              0.01   
4                                           0              3.26   

                 City  
0  New York, New York  
1  New York, New York  
2  New York, New York  
3  New York, New York  
4  New York, New York  

[5 rows x 75 columns]

In [6]:
amenities = []
for s in new_york_listings.amenities.values: # s is a str of what looks like a Python list
    ls = re.findall(r'"(.*?)"', s) # ls is a list of the amenities listed in s
    amenities.extend(ls)
amenities = set(amenities)
amenities # set containing all unique 

{' THERE IS A CELLING FAN AN A PORTABLE AC IN THE BEDROOM conditioner',
 ' linens',
 ' toiletries',
 ', ',
 '100\\',
 '120\\',
 '12\\',
 '2 burner induction cooktop.  We provide pots and pans. induction stove',
 '20\\',
 '24-hour fitness center',
 '24\\',
 '26\\',
 '30\\',
 '32\\',
 '36\\',
 '37\\',
 '39\\',
 '40\\',
 '42\\',
 '43\\',
 '45\\',
 '46\\',
 '47\\',
 '48\\',
 '49\\',
 '50\\',
 '51\\',
 '52\\',
 '54\\',
 '55\\',
 '58\\',
 '60\\',
 '62\\',
 '64\\',
 '65\\',
 '72\\',
 '75\\',
 '84\\',
 '85\\',
 'Acca Kappa toiletries',
 'Aesop, Dr. Bronners body soap',
 'Air conditioning',
 'Airport shuttle',
 'Alexa  sound system with Bluetooth and aux',
 'Alexa Bluetooth sound system',
 'Alexa sound system with Bluetooth and aux',
 'All inclusive',
 'Amazon Alexa sound system',
 'Americana refrigerator',
 'Amika conditioner',
 'Amika shampoo',
 'Anker  Bluetooth sound system',
 'Appelles toiletries',
 'Apple AirPlay sound system',
 'Apple HomePods Bluetooth sound system',
 'Asprey from the R

In [7]:
new_york_listings['amenities'] = new_york_listings['amenities'].apply(lambda x: re.findall(r'"(.*?)"', x))

In [8]:
def get_amenities_cols(df, amenities):
    """Input df and list of amenities, return df with binary column for each amenity in amenities"""
    for amen in list(amenities):
        df[amen] = df['amenities'].apply(lambda x: 1 if amen in x else 0)
    return df

To Do:
Find most common amenities from successful and very unsuccessful units. This means getting top and bottom 15(ish)% of listings (based on price), and get a count of each amenity listing for those. Maybe make a bar chart with a bar for top earners and a bar for low earners?

In [9]:
new_york_listings['price'] = new_york_listings['price'].apply(lambda x: float(re.sub(r'[^\d.]', '', x)))
new_york_listings['price']
# re.sub('[^\d.]*', '', '     $$$$1234567890.98')

0        150.0
1         76.0
2         60.0
3        175.0
4         79.0
         ...  
36900     46.0
36901     98.0
36902    145.0
36903    123.0
36904    299.0
Name: price, Length: 36905, dtype: float64

In [10]:
# new_york_listings.columns

In [11]:
def get_top_or_bot(df, ratio, bottom=True):
    """Sort df by price and get top/bottom ratio (between 0-1)"""
    sorted_df = df.sort_values('price', ascending=bottom)
    return sorted_df.head(int(len(df)*ratio))

In [12]:
# ratio = 0.25
# top_df = get_top_or_bot(new_york_listings, ratio, bottom=False)
# # top_count_df = top_df.groupby('price').agg('count')
# # top_count_df

In [13]:
# bottom_df = get_top_or_bot(new_york_listings, ratio, bottom=True)
# # bottom_count_df = bottom_df.groupby('price').agg('count')
# # bottom_count_df

In [92]:
def compare_amenities(df, ratio):
    top_df = get_top_or_bot(new_york_listings, ratio, bottom=False)
    bottom_df = get_top_or_bot(new_york_listings, ratio, bottom=True)
    top_amen_cnt = []
    bot_amen_cnt = []

    for amen in amenities:
        top_cnt = 0
        for ls in top_df.amenities.values:
            if amen in ls:
                top_cnt += 1
        top_amen_cnt.append(top_cnt)

        bot_cnt = 0
        for ls in bottom_df.amenities.values:
            if amen in ls:
                bot_cnt += 1
        bot_amen_cnt.append(bot_cnt)

    amen_cnt_df = pd.DataFrame(index=amenities)
    amen_cnt_df['top'] = top_amen_cnt
    amen_cnt_df['bot'] = bot_amen_cnt
    amen_cnt_df = amen_cnt_df[(amen_cnt_df['top'] > 0)|(amen_cnt_df['bot'] > 0)]
    amen_cnt_df['dif'] = amen_cnt_df['top'] - amen_cnt_df['bot']
    return alt.Chart(amen_cnt_df.reset_index(), height=300, width=850).mark_bar().encode(
        x = alt.X('index', sort="-y", axis=alt.Axis(title='Amenities', labels=False)),
        y = alt.Y('dif', axis=alt.Axis(title='Difference in Amenity Count')),
        tooltip = 'index'
    )

In [15]:
# top_amen_cnt = []
# bot_amen_cnt = []

# for amen in amenities:
#     top_cnt = 0
#     for ls in top_df.amenities.values:
#         if amen in ls:
#             top_cnt += 1
#     top_amen_cnt.append(top_cnt)
    
#     bot_cnt = 0
#     for ls in bottom_df.amenities.values:
#         if amen in ls:
#             bot_cnt += 1
#     bot_amen_cnt.append(bot_cnt)

# amen_cnt_df = pd.DataFrame(index=amenities)
# amen_cnt_df['top'] = top_amen_cnt
# amen_cnt_df['bot'] = bot_amen_cnt
# amen_cnt_df = amen_cnt_df[(amen_cnt_df['top'] > 0)|(amen_cnt_df['bot'] > 0)]

In [16]:
# amen_cnt_df['dif'] = amen_cnt_df['top'] - amen_cnt_df['bot']

# alt.Chart(amen_cnt_df.reset_index()).mark_bar().encode(
#     x = alt.X('index', sort="-y"),
#     y = alt.Y('dif')
# )


In [17]:
compare_amenities(new_york_listings, 0.25)

alt.Chart(...)

In [18]:
compare_amenities(new_york_listings, 0.005)

alt.Chart(...)

In [19]:
compare_amenities(new_york_listings, 0.5)

alt.Chart(...)

In [20]:
amen_cols = ['TV', 'Dryer', 'Shampoo', 'Gym', 'Private entrance', 'Patio or balcony', 
             'Pool', 'Laundromat nearby', 'Host greets you', 'Free street parking', 
             'Lock on bedroom door']
# amen_cols = ['TV', 'Dryer', 'Shampoo', 'Gym', 'Pool']


# 'Paid parking off premises',
# 'Cleaning before checkout',
# 'Indoor fireplace',
# 'Security cameras on property',
# exp_df.columns

In [21]:
exp_df = get_amenities_cols(new_york_listings, amen_cols)
to_rename = {'Private entrance':'private_entrance', 'Patio or balcony':'patio_or_balcony', 
            'Laundromat nearby':'laundromat_nearby', 'Host greets you':'host_greets_you', 
             'Free street parking':'free_street_parking', 'Lock on bedroom door':'lock_on_bedroom_door'}
amen_cols = ['TV', 'Dryer', 'Shampoo', 'Gym', 'private_entrance', 'patio_or_balcony', 
             'Pool', 'laundromat_nearby', 'host_greets_you', 'free_street_parking', 
             'lock_on_bedroom_door']
exp_df = exp_df.rename(columns=to_rename)
exp_df

id                            listing_url       scrape_id  \
0          2595      https://www.airbnb.com/rooms/2595  20210407173759   
1          3831      https://www.airbnb.com/rooms/3831  20210407173759   
2          5121      https://www.airbnb.com/rooms/5121  20210407173759   
3          5136      https://www.airbnb.com/rooms/5136  20210407173759   
4          5178      https://www.airbnb.com/rooms/5178  20210407173759   
...         ...                                    ...             ...   
36900  49088566  https://www.airbnb.com/rooms/49088566  20210407173759   
36901  49089512  https://www.airbnb.com/rooms/49089512  20210407173759   
36902  49089765  https://www.airbnb.com/rooms/49089765  20210407173759   
36903  49091581  https://www.airbnb.com/rooms/49091581  20210407173759   
36904  49092532  https://www.airbnb.com/rooms/49092532  20210407173759   

      last_scraped                                               name  \
0       2021-04-09                              Skylit Midtown Castle   
1       2021-04-12  Whole flr w/private bdrm, bath & kitchen(pls r...   
2       2021-04-09                                    BlissArtsSpace!   
3       2021-04-10           Spacious Brooklyn Duplex, Patio + Garden   
4       2021-04-12                   Large Furnished Room Near B'way　   
...            ...                                                ...   
36900   2021-04-10                  Amenity-loaded Single Room in UWS   
36901   2021-04-09       Unfurnished 2BR Apartment in Upper East Side   
36902   2021-04-09             Beautiful Renovated  Sunny Jr. One Bed   
36903   2021-04-09       Furnished Bright Apartment in Hell's Kitchen   
36904   2021-04-09  Panoramic city views!  2/3 bdrms private terra...   

                                             description  \
0      Beautiful, spacious skylit studio in the heart...   
1      Enjoy 500 s.f. top floor in 1899 brownstone, w...   
2      <b>The space</b><br />HELLO EVERYONE AND THANK...   
3      We welcome you to stay in our lovely 2 br dupl...   
4      Please don’t expect the luxury here just a bas...   
...                                                  ...   
36900  SharedEasy is the only community-oriented Coli...   
36901  2 bedroom apartment in Upper East Side.<br /><...   
36902  Beautifully recently renovated and furnished J...   
36903  1 bedroom apartment in Hell's Kitchen.<br />Mo...   
36904  You’ll be right in the heart of everything at ...   

                                   neighborhood_overview  \
0      Centrally located in the heart of Manhattan ju...   
1      Just the right mix of urban center and local n...   
2                                                    NaN   
3                                                    NaN   
4        Theater district, many restaurants around here.   
...                                                  ...   
36900  Morningside Heights is bursting at the seams w...   
36901                                                NaN   
36902                                                NaN   
36903                                                NaN   
36904  Best neighborhood in NYC !! You have easy walk...   

                                             picture_url    host_id  \
0      https://a0.muscache.com/pictures/f0813a11-40b2...       2845   
1      https://a0.muscache.com/pictures/e49999c2-9fd5...       4869   
2      https://a0.muscache.com/pictures/2090980c-b68e...       7356   
3      https://a0.muscache.com/pictures/a3f7ace8-e7a9...       7378   
4      https://a0.muscache.com/pictures/12065/f070997...       8967   
...                                                  ...        ...   
36900  https://a0.muscache.com/pictures/9eeedafc-e693...  334376746   
36901  https://a0.muscache.com/pictures/f80f622a-3194...  305240193   
36902  https://a0.muscache.com/pictures/miso/Hosting-...  387269117   
36903  https://a0.muscache.com/pictures/8f730543-a68b...  305240193   
36904  https://a0.muscac

In [22]:
# exp_df['neighbourhood']

In [23]:
formula = "price ~ host_id + host_listings_count + host_has_profile_pic + host_identity_verified + bedrooms + beds + availability_60 + number_of_reviews_l30d + review_scores_rating + "
for count, el in enumerate(amen_cols):   
    if count < len(amen_cols) - 1:
        formula = formula + el + "+"
    else:
        formula = formula + el
reg = smf.ols(formula=formula, data=exp_df).fit()
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.094
Model:                            OLS   Adj. R-squared:                  0.093
Method:                 Least Squares   F-statistic:                     119.2
Date:                Mon, 23 Aug 2021   Prob (F-statistic):               0.00
Time:                        16:37:43   Log-Likelihood:            -1.6080e+05
No. Observations:               22982   AIC:                         3.216e+05
Df Residuals:                   22961   BIC:                         3.218e+05
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                     -27.7216     40.898     -0.678      0.498    -107.885      52.441
host_has_profile_pic[T.t]     -20.0377     36.792     -0.545      0.586     -92.152      52.077
host_identity_verified[T.t]    -2.6883      4.522     -0.595      0.552     -11.551       6.174
host_id                     -7.662e-09   1.82e-08     -0.421      0.674   -4.34e-08    2.81e-08
host_listings_count             0.0379      0.022      1.691      0.091      -0.006       0.082
bedrooms                       80.0360      3.711     21.565      0.000      72.761      87.311
beds                           10.2004      2.166      4.710      0.000       5.956      14.445
availability_60                 0.6598      0.074      8.863      0.000       0.514       0.806
number_of_reviews_l30d         -4.0620      1.309     -3.104      0.002      -6.627      -1.497
review_scores_rating            0.5010      0.186      2.699      0.007       0.137       0.865
TV                             22.7367      3.956      5.747      0.000      14.983      30.491
Dryer                          17.8141      3.805      4.682      0.000      10.357      25.271
Shampoo                        21.9310      3.911      5.608      0.000      14.266      29.596
Gym                            52.7382      7.059      7.471      0.000      38.903      66.574
private_entrance               -1.6516      4.421     -0.374      0.709     -10.316       7.013
patio_or_balcony               24.2146      5.641      4.292      0.000      13.158      35.272
Pool                           77.5109     17.452      4.441      0.000      43.304     111.718
laundromat_nearby              -8.1344      8.028     -1.013      0.311     -23.870       7.602
host_greets_you                 5.0385      4.919      1.024      0.306      -4.604      14.681
free_street_parking           -44.6728      3.876    -11.527      0.000     -52.269     -37.076
lock_on_bedroom_door          -46.4449      4.228    -10.986      0.000     -54.732     -38.158
==============================================================================
Omnibus:                    56757.421   Durbin-Watson:                   1.834
Prob(Omnibus):                  0.000   Jarque-Bera (JB):        800068891.502
Skew:                          26.672   Prob(JB):                         0.00
Kurtosis:                     915.504   Cond. No.                     4.14e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.14e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The above regression was decently successful. Most importantly, it proved a statistically significant effect for a number of features, so we may move forward with using these features in our predictive analysis. However, there is a problem of multicollinearity indicated by the large condition number (?). 

We also want to pivot to revenue as the dependent variable, since it incorporates both nightly price and demand (availability/days rented out) over the next (blank) days. Let's start working on the time-series data since that is what we want to do the actual regression based off of. We need to figure out how to get days booked over the past 90 days as a column in time-series.

In [24]:
#Google Cloud
from sqlalchemy import create_engine
import psycopg2

conn = psycopg2.connect(host="34.138.230.114", port = "5432", database="postgres", user="postgres", password="Sparty2723!")

# Create a cursor object
cur = conn.cursor()

#dialect+driver://username:password@host:port/database
alchemyEngine           = create_engine('postgresql+psycopg2://postgres:Sparty2723!@34.138.230.114:5432/postgres');
postgreSQLConnection    = alchemyEngine.connect();

conn = postgreSQLConnection.connect()

sql_command = "SELECT * FROM nyc_calendar_data LIMIT 10;"
cur.execute(sql_command)

cur.fetchone()

(0,
 55982,
 datetime.datetime(2021, 8, 4, 0, 0),
 False,
 85.0,
 85.0,
 3.0,
 18.0,
 8,
 2021)

In [25]:
pd.DataFrame(cur.fetchall())

0     1          2      3      4      5     6       7  8     9
0  1  2595 2021-08-04   True  150.0  150.0  30.0  1125.0  8  2021
1  2  2595 2021-08-05   True  150.0  150.0  30.0  1125.0  8  2021
2  3  2595 2021-08-06   True  150.0  150.0  30.0  1125.0  8  2021
3  4  2595 2021-08-07  False  150.0  150.0  30.0  1125.0  8  2021
4  5  2595 2021-08-08  False  150.0  150.0  30.0  1125.0  8  2021
5  6  2595 2021-08-09  False  150.0  150.0  30.0  1125.0  8  2021
6  7  2595 2021-08-10  False  150.0  150.0  30.0  1125.0  8  2021
7  8  2595 2021-08-11  False  150.0  150.0  30.0  1125.0  8  2021
8  9  2595 2021-08-12  False  150.0  150.0  30.0  1125.0  8  2021

In [26]:
sql_command = "SELECT * FROM nyc_calendar_data;"
cur.execute(sql_command)
nyc_calendar_data_df = pd.DataFrame(cur.fetchall())

In [27]:
nyc_calendar_data_df = nyc_calendar_data_df.rename(columns={0:'index', 1:'listing_id', 2:'date', 3:'available', 
                                                            4:'price', 5:'adjusted_price', 6:'minimum_nights', 
                                                            7:'maximum_nights', 8:'month', 9:'year'})
nyc_calendar_data_df

index  listing_id       date  available  price  adjusted_price  \
0                0       55982 2021-08-04      False   85.0            85.0   
1                1        2595 2021-08-04       True  150.0           150.0   
2                2        2595 2021-08-05       True  150.0           150.0   
3                3        2595 2021-08-06       True  150.0           150.0   
4                4        2595 2021-08-07      False  150.0           150.0   
...            ...         ...        ...        ...    ...             ...   
13397118  13397118    51447515 2022-07-30       True  255.0           255.0   
13397119  13397119    51447515 2022-07-31       True  255.0           255.0   
13397120  13397120    51447515 2022-08-01       True  255.0           255.0   
13397121  13397121    51447515 2022-08-02       True  255.0           255.0   
13397122  13397122    51447515 2022-08-03       True  255.0           255.0   

          minimum_nights  maximum_nights  month  year  
0                    3.0            18.0      8  2021  
1                   30.0          1125.0      8  2021  
2                   30.0          1125.0      8  2021  
3                   30.0          1125.0      8  2021  
4                   30.0          1125.0      8  2021  
...                  ...             ...    ...   ...  
13397118            31.0          1125.0      7  2022  
13397119            31.0          1125.0      7  2022  
13397120            31.0          1125.0      8  2022  
13397121            31.0          1125.0      8  2022  
13397122            31.0          1125.0      8  2022  

[13397123 rows x 10 columns]

In [ ]:
new_york_listings = new_york_listings.rename(columns=to_rename)
new_york_listings = new_york_listings.set_index('id') # 'id' must be the index for the below function to work 

In [35]:
# new_york_listings['host_greets_you'].loc[2595]

In [36]:
# new_york_listings = new_york_listings.set_index('id')
def assign_amen(df, amen, listing_id):
    try:
        return df[amen].loc[listing_id]
    except:
        return np.nan
amen_cols = ["host_listings_count", "host_has_profile_pic", "host_identity_verified", "bedrooms", "beds", 
             "availability_60", "number_of_reviews_l30d", "review_scores_rating", 'TV', 'Dryer', 'Shampoo', 
             'Gym', 'private_entrance', 'patio_or_balcony', 'Pool', 'laundromat_nearby', 'host_greets_you', 
             'free_street_parking', 'lock_on_bedroom_door']


for amen in amen_cols:
    print(amen)
    nyc_calendar_data_df[amen] = nyc_calendar_data_df['listing_id'].apply(lambda x: assign_amen(new_york_listings, amen, x))
nyc_calendar_data_df = nyc_calendar_data_df.rename(columns=to_rename)
nyc_calendar_data_df

host_listings_count
host_has_profile_pic
host_identity_verified
bedrooms
beds
availability_60
number_of_reviews_l30d
review_scores_rating
TV
Dryer
Shampoo
Gym
private_entrance
patio_or_balcony
Pool
laundromat_nearby
host_greets_you
free_street_parking
lock_on_bedroom_door


index  listing_id       date  available  price  adjusted_price  \
0                0       55982 2021-08-04      False   85.0            85.0   
1                1        2595 2021-08-04       True  150.0           150.0   
2                2        2595 2021-08-05       True  150.0           150.0   
3                3        2595 2021-08-06       True  150.0           150.0   
4                4        2595 2021-08-07      False  150.0           150.0   
...            ...         ...        ...        ...    ...             ...   
13397118  13397118    51447515 2022-07-30       True  255.0           255.0   
13397119  13397119    51447515 2022-07-31       True  255.0           255.0   
13397120  13397120    51447515 2022-08-01       True  255.0           255.0   
13397121  13397121    51447515 2022-08-02       True  255.0           255.0   
13397122  13397122    51447515 2022-08-03       True  255.0           255.0   

          minimum_nights  maximum_nights  month  year  ...  Dryer Shampoo  \
0                    3.0            18.0      8  2021  ...    0.0     1.0   
1                   30.0          1125.0      8  2021  ...    0.0     0.0   
2                   30.0          1125.0      8  2021  ...    0.0     0.0   
3                   30.0          1125.0      8  2021  ...    0.0     0.0   
4                   30.0          1125.0      8  2021  ...    0.0     0.0   
...                  ...             ...    ...   ...  ...    ...     ...   
13397118            31.0          1125.0      7  2022  ...    NaN     NaN   
13397119            31.0          1125.0      7  2022  ...    NaN     NaN   
13397120            31.0          1125.0      8  2022  ...    NaN     NaN   
13397121            31.0          1125.0      8  2022  ...    NaN     NaN   
13397122            31.0          1125.0      8  2022  ...    NaN     NaN   

          Gym  private_entrance  patio_or_balcony  Pool  laundromat_nearby  \
0         0.0               0.0               0.0   0.0                0.0   
1         0.0               0.0               0.0   0.0                0.0   
2         0.0               0.0               0.0   0.0                0.0   
3         0.0               0.0               0.0   0.0                0.0   
4         0.0               0.0               0.0   0.0                0.0   
...       ...               ...               ...   ...                ...   
13397118  NaN               NaN               NaN   NaN                NaN   
13397119  NaN               NaN               NaN   NaN                NaN   
13397120  NaN               NaN               NaN   NaN                NaN   
13397121  NaN               NaN               NaN   NaN                NaN   
13397122  NaN               NaN               NaN   NaN                NaN   

          host_greets_you  free_street_parking  lock_on_bedroom_door  
0                     1.0                  1.0                   0.0  
1                     0.0                  1.0                   0.0  
2                     0.0                  1.0                   0.0  
3                     0.0                  1.0                   0.0  
4                     0.0                  1.0                   0.0  
...                   ...                  ...                   ...  
13397118              NaN                  NaN                   NaN  
13397119              NaN                  NaN                   NaN  
13397120              NaN                  NaN                   NaN  
13397121              NaN                  NaN                   NaN  
13397122              NaN                  NaN                   NaN  

[13397123 rows x 29 columns]

In [37]:
new_york_listings.isna().sum() # uh-oh! basically all NaNs for some reason in the added columns. Took ~35 mins

listing_url                0
scrape_id                  0
last_scraped               0
name                      13
description             1268
                        ... 
Pool                       0
laundromat_nearby          0
host_greets_you            0
free_street_parking        0
lock_on_bedroom_door       0
Length: 85, dtype: int64

In [38]:
# to_rename = {'Private entrance':'private_entrance', 'Patio or balcony':'patio_or_balcony', 
#             'Laundromat nearby':'laundromat_nearby', 'Host greets you':'host_greets_you', 
#              'Free street parking':'free_street_parking', 'Lock on bedroom door':'lock_on_bedroom_door'}
# nyc_calendar_data_df = nyc_calendar_data_df.rename(columns=to_rename) # already renamed columns earlier
dropped_df = nyc_calendar_data_df.dropna() # about 86% the size of nyc_calendar_data_df
dropped_df

index  listing_id       date  available  price  adjusted_price  \
0                0       55982 2021-08-04      False   85.0            85.0   
97              97       15341 2021-08-04      False  192.0           192.0   
98              98       15341 2021-08-05      False  192.0           192.0   
99              99       15341 2021-08-06      False  190.0           190.0   
100            100       15341 2021-08-07      False  190.0           190.0   
...            ...         ...        ...        ...    ...             ...   
12142133  12142133    49003697 2022-07-30       True  150.0           150.0   
12142134  12142134    49003697 2022-07-31       True  150.0           150.0   
12142135  12142135    49003697 2022-08-01       True  150.0           150.0   
12142136  12142136    49003697 2022-08-02       True  150.0           150.0   
12142137  12142137    49003697 2022-08-03       True  150.0           150.0   

          minimum_nights  maximum_nights  month  year  ...  Dryer Shampoo  \
0                    3.0            18.0      8  2021  ...    0.0     1.0   
97                  30.0          1125.0      8  2021  ...    0.0     0.0   
98                  30.0          1125.0      8  2021  ...    0.0     0.0   
99                  30.0          1125.0      8  2021  ...    0.0     0.0   
100                 30.0          1125.0      8  2021  ...    0.0     0.0   
...                  ...             ...    ...   ...  ...    ...     ...   
12142133             3.0          1125.0      7  2022  ...    0.0     0.0   
12142134             3.0          1125.0      7  2022  ...    0.0     0.0   
12142135             3.0          1125.0      8  2022  ...    0.0     0.0   
12142136             3.0          1125.0      8  2022  ...    0.0     0.0   
12142137             3.0          1125.0      8  2022  ...    0.0     0.0   

          Gym  private_entrance  patio_or_balcony  Pool  laundromat_nearby  \
0         0.0               0.0               0.0   0.0                0.0   
97        0.0               0.0               1.0   0.0                1.0   
98        0.0               0.0               1.0   0.0                1.0   
99        0.0               0.0               1.0   0.0                1.0   
100       0.0               0.0               1.0   0.0                1.0   
...       ...               ...               ...   ...                ...   
12142133  0.0               0.0               0.0   0.0                0.0   
12142134  0.0               0.0               0.0   0.0                0.0   
12142135  0.0               0.0               0.0   0.0                0.0   
12142136  0.0               0.0               0.0   0.0                0.0   
12142137  0.0               0.0               0.0   0.0                0.0   

          host_greets_you  free_street_parking  lock_on_bedroom_door  
0                     1.0                  1.0                   0.0  
97                    0.0                  0.0                   0.0  
98                    0.0                  0.0                   0.0  
99                    0.0                  0.0                   0.0  
100                   0.0                  0.0                   0.0  
...                   ...                  ...                   ...  
12142133              0.0                  0.0                   0.0  
12142134              0.0                  0.0                   0.0  
12142135              0.0                  0.0                   0.0  
12142136              0.0                  0.0                   0.0  
12142137              0.0                  0.0                   0.0  

[7328553 rows x 29 columns]

In [39]:
#new_york_listings = new_york_listings.set_index('id')
# nyc_calendar_data_df[~nyc_calendar_data_df.set_index('listing_id').isin(new_york_listings)].dropna()

# nyc_calendar_data_df.set_index(['listing_id', 'date'])

In [40]:
outliers = set(nyc_calendar_data_df.listing_id.values) - set(new_york_listings.index)
len(outliers)

4950

There are 4950 listings in the nyc_calendar_df that are not included in the new_york_listings df. Thus we will simply drop these listings from our df, resulting in the dropped_df above. Additionally, there are some missing review scores, resulting in NaN values that were subsequently dropped. The remaining df is approximately 54% the size of the initial df, with upwards of 7M rows. 

Because the price of AirBnB units likely is dependent on the time of year, we would usually want to do a seasonal decomposition (using statsmodels.tsa.seasonal seasonal_decompose, DecomposeResult). Since we only have 1 year of data, however, we do not have enough cycles (years) of data to complete such a decomposition. Thus, we will simply treat month, year, and day as their own columns containing integers. In future experimentation, it would be ideal to have data spanning the past few years. This could make for much more robust models for predicting prices and when units would be rented out.

**Thought about regressing to find revenue. Let's initially regress to find price. Then try grouping by month, averaging the price and summing the available (must change to binary first). These can then be multiplied to get monthly revenue.**

In [41]:
dropped_df['day'] = dropped_df['date'].dt.day
dropped_df['weekday'] = dropped_df['date'].dt.weekday # include day of the week

dropped_df

<ipython-input-41-714c3397886b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dropped_df['day'] = dropped_df['date'].dt.day
<ipython-input-41-714c3397886b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dropped_df['weekday'] = dropped_df['date'].dt.weekday # include day of the week


index  listing_id       date  available  price  adjusted_price  \
0                0       55982 2021-08-04      False   85.0            85.0   
97              97       15341 2021-08-04      False  192.0           192.0   
98              98       15341 2021-08-05      False  192.0           192.0   
99              99       15341 2021-08-06      False  190.0           190.0   
100            100       15341 2021-08-07      False  190.0           190.0   
...            ...         ...        ...        ...    ...             ...   
12142133  12142133    49003697 2022-07-30       True  150.0           150.0   
12142134  12142134    49003697 2022-07-31       True  150.0           150.0   
12142135  12142135    49003697 2022-08-01       True  150.0           150.0   
12142136  12142136    49003697 2022-08-02       True  150.0           150.0   
12142137  12142137    49003697 2022-08-03       True  150.0           150.0   

          minimum_nights  maximum_nights  month  year  ...  Gym  \
0                    3.0            18.0      8  2021  ...  0.0   
97                  30.0          1125.0      8  2021  ...  0.0   
98                  30.0          1125.0      8  2021  ...  0.0   
99                  30.0          1125.0      8  2021  ...  0.0   
100                 30.0          1125.0      8  2021  ...  0.0   
...                  ...             ...    ...   ...  ...  ...   
12142133             3.0          1125.0      7  2022  ...  0.0   
12142134             3.0          1125.0      7  2022  ...  0.0   
12142135             3.0          1125.0      8  2022  ...  0.0   
12142136             3.0          1125.0      8  2022  ...  0.0   
12142137             3.0          1125.0      8  2022  ...  0.0   

         private_entrance patio_or_balcony  Pool  laundromat_nearby  \
0                     0.0              0.0   0.0                0.0   
97                    0.0              1.0   0.0                1.0   
98                    0.0              1.0   0.0                1.0   
99                    0.0              1.0   0.0                1.0   
100                   0.0              1.0   0.0                1.0   
...                   ...              ...   ...                ...   
12142133              0.0              0.0   0.0                0.0   
12142134              0.0              0.0   0.0                0.0   
12142135              0.0              0.0   0.0                0.0   
12142136              0.0              0.0   0.0                0.0   
12142137              0.0              0.0   0.0                0.0   

          host_greets_you  free_street_parking  lock_on_bedroom_door  day  \
0                     1.0                  1.0                   0.0    4   
97                    0.0                  0.0                   0.0    4   
98                    0.0                  0.0                   0.0    5   
99                    0.0                  0.0                   0.0    6   
100                   0.0                  0.0                   0.0    7   
...                   ...                  ...                   ...  ...   
12142133              0.0                  0.0                   0.0   30   
12142134              0.0                  0.0                   0.0   31   
12142135              0.0                  0.0                   0.0    1   
12142136              0.0                  0.0                   0.0    2   
12142137              0.0                  0.0                   0.0    3   

          weekday  
0               2  
97              2  
98              3  
99              4  
100             5  
...           ...  
12142133        5  
12142134        6  
12142135        0  
12142136        1  
12142137        2  

[7328553 rows x 31 columns]

In [42]:
reg1_df = dropped_df.drop(columns=['index','date','available','adjusted_price'])
reg1_df

listing_id  price  minimum_nights  maximum_nights  month  year  \
0              55982   85.0             3.0            18.0      8  2021   
97             15341  192.0            30.0          1125.0      8  2021   
98             15341  192.0            30.0          1125.0      8  2021   
99             15341  190.0            30.0          1125.0      8  2021   
100            15341  190.0            30.0          1125.0      8  2021   
...              ...    ...             ...             ...    ...   ...   
12142133    49003697  150.0             3.0          1125.0      7  2022   
12142134    49003697  150.0             3.0          1125.0      7  2022   
12142135    49003697  150.0             3.0          1125.0      8  2022   
12142136    49003697  150.0             3.0          1125.0      8  2022   
12142137    49003697  150.0             3.0          1125.0      8  2022   

          host_listings_count host_has_profile_pic host_identity_verified  \
0                         1.0                    t                      t   
97                        1.0                    t                      t   
98                        1.0                    t                      t   
99                        1.0                    t                      t   
100                       1.0                    t                      t   
...                       ...                  ...                    ...   
12142133                  0.0                    t                      t   
12142134                  0.0                    t                      t   
12142135                  0.0                    t                      t   
12142136                  0.0                    t                      t   
12142137                  0.0                    t                      t   

          bedrooms  ...  Gym  private_entrance  patio_or_balcony  Pool  \
0              1.0  ...  0.0               0.0               0.0   0.0   
97             1.0  ...  0.0               0.0               1.0   0.0   
98             1.0  ...  0.0               0.0               1.0   0.0   
99             1.0  ...  0.0               0.0               1.0   0.0   
100            1.0  ...  0.0               0.0               1.0   0.0   
...            ...  ...  ...               ...               ...   ...   
12142133       1.0  ...  0.0               0.0               0.0   0.0   
12142134       1.0  ...  0.0               0.0               0.0   0.0   
12142135       1.0  ...  0.0               0.0               0.0   0.0   
12142136       1.0  ...  0.0               0.0               0.0   0.0   
12142137       1.0  ...  0.0               0.0               0.0   0.0   

          laundromat_nearby  host_greets_you  free_street_parking  \
0                       0.0              1.0                  1.0   
97                      1.0              0.0                  0.0   
98                      1.0              0.0                  0.0   
99                      1.0              0.0                  0.0   
100                     1.0              0.0                  0.0   
...                     ...              ...                  ...   
12142133                0.0              0.0                  0.0   
12142134                0.0              0.0                  0.0   
12142135                0.0              0.0                  0.0   
12142136                0.0              0.0                  0.0   
12142137                0.0              0.0                  0.0   

          lock_on_bedroom_door  day  weekday  
0                          0.0    4        2  
97                         0.0    4        2  
98                         0.0    5        3  
99                         0.0    6        4  
100                        0.0    7        5  
...                        ...  ...      ...  
12142133                   0.0   30        5  
12142134                   0.0   31        6  
12142135                   0.0 

In [43]:
reg1_df.columns

Index(['listing_id', 'price', 'minimum_nights', 'maximum_nights', 'month',
       'year', 'host_listings_count', 'host_has_profile_pic',
       'host_identity_verified', 'bedrooms', 'beds', 'availability_60',
       'number_of_reviews_l30d', 'review_scores_rating', 'TV', 'Dryer',
       'Shampoo', 'Gym', 'private_entrance', 'patio_or_balcony', 'Pool',
       'laundromat_nearby', 'host_greets_you', 'free_street_parking',
       'lock_on_bedroom_door', 'day', 'weekday'],
      dtype='object')

In [44]:
formula = "price ~ minimum_nights + maximum_nights + month + year + day + weekday + host_listings_count + \
           host_has_profile_pic + host_identity_verified + bedrooms + beds + availability_60 + \
           number_of_reviews_l30d + review_scores_rating + TV + Dryer + Shampoo + Gym + Pool + private_entrance + \
           patio_or_balcony + laundromat_nearby + host_greets_you + free_street_parking + lock_on_bedroom_door"

reg1 = smf.ols(formula=formula, data=reg1_df).fit()
reg1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.093
Model:                            OLS   Adj. R-squared:                  0.093
Method:                 Least Squares   F-statistic:                 3.013e+04
Date:                Mon, 23 Aug 2021   Prob (F-statistic):               0.00
Time:                        17:48:15   Log-Likelihood:            -5.1965e+07
No. Observations:             7328553   AIC:                         1.039e+08
Df Residuals:                 7328527   BIC:                         1.039e+08
Df Model:                          25                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                   -9609.3256    841.273    -11.422      0.000   -1.13e+04   -7960.461
host_has_profile_pic[T.t]     -17.5432      2.273     -7.717      0.000     -21.999     -13.087
host_identity_verified[T.t]     5.0324      0.276     18.251      0.000       4.492       5.573
minimum_nights                 -0.0462      0.003    -13.636      0.000      -0.053      -0.040
maximum_nights               1.049e-07   1.53e-09     68.460      0.000    1.02e-07    1.08e-07
month                           0.8349      0.059     14.076      0.000       0.719       0.951
year                            4.7529      0.416     11.426      0.000       3.938       5.568
day                             0.0176      0.012      1.445      0.149      -0.006       0.042
weekday                         0.4418      0.054      8.223      0.000       0.337       0.547
host_listings_count             0.0375      0.002     21.338      0.000       0.034       0.041
bedrooms                       78.8251      0.230    342.802      0.000      78.374      79.276
beds                           14.6626      0.135    108.304      0.000      14.397      14.928
availability_60                 0.9400      0.005    202.838      0.000       0.931       0.949
number_of_reviews_l30d         -0.2257      0.079     -2.869      0.004      -0.380      -0.072
review_scores_rating           -0.0251      0.012     -2.071      0.038      -0.049      -0.001
TV                             23.5290      0.243     96.677      0.000      23.052      24.006
Dryer                          18.8892      0.234     80.645      0.000      18.430      19.348
Shampoo                        30.7343      0.240    127.861      0.000      30.263      31.205
Gym                            57.9830      0.431    134.561      0.000      57.138      58.828
Pool                          100.8370      1.093     92.296      0.000      98.696     102.978
private_entrance                6.5046      0.272     23.906      0.000       5.971       7.038
patio_or_balcony               29.6777      0.349     84.919      0.000      28.993      30.363
laundromat_nearby               1.0299      0.498      2.068      0.039       0.054       2.006
host_greets_you                 4.9316      0.302     16.326      0.000       4.340       5.524
free_street_parking           -52.2150      0.240   -217.620      0.000     -52.685     -51.745
lock_on_bedroom_door          -46.7877      0.259   -180.796      0.000     -47.295     -46.280
==============================================================================
Omnibus:                 17653470.882   Durbin-Watson:                   0.048
Prob(Omnibus):                  0.000   Jarque-Bera (JB):     198498559973.773
Skew:                          25.249   Prob(JB):             

The first regression on calendar data (reg1) shows that every chosen feature, aside from day of the month (day), is statistically significant with p-values of zero at 3 decimal places. However, the R-squared value is still quite low at only 0.093. Let's see if we can get that to go up by adding in more external features from the New York City open data. These features are by borough, so we must also add in which borough each is listing resides in. 

**The data from the NYC open data source is outdated, lasting from 2010 through 2018. The calendar data is from 2021 to 2022. We will be using the 2017 to 2018 data as though it were the appropriate years for the calendar data since it is the most contemporary data that we have. Assuming there are lasting trends from 2017 to now, perhaps these values can provide serviceable estimates. But the main point of this regression is to prove the ability of this data to predict price of AirBnB units given certain features of the unit and the surrounding area. Given more accurate data, the model could perform better.**

In [45]:
external_features_df = pd.read_csv('assets/Literally_Everything.csv')
external_features_df = external_features_df[(external_features_df['YEAR']==2017)|(external_features_df['YEAR']==2018)]
external_features_df.sample(5)

Borough    YEAR  Total housing units  Occupied housing units  \
7       Bronx  2017.0             528851.0                503985.0   
10     Queens  2017.0             857250.0                769258.0   
1       Bronx  2018.0             532509.0                507370.0   
4      Queens  2018.0             865809.0                788110.0   
9   Manhattan  2017.0             886384.0                764218.0   

    Vacant housing units  1-unit, detached  1-unit, attached   2 units  \
7                24866.0           31166.0           27320.0   42804.0   
10               87992.0          156959.0           72776.0  177407.0   
1                25139.0           32229.0           22889.0   47809.0   
4                77699.0          163013.0           71163.0  177611.0   
9               122166.0           11112.0            7629.0    8015.0   

    3 or 4 units  5 to 9 units  ...      Population 25 years and over  \
7        43653.0       19731.0  ...                          953271.0   
10       90970.0       47730.0  ...                         1695302.0   
1        40522.0       20169.0  ...                          934391.0   
4        87785.0       51055.0  ...                         1646542.0   
9        17953.0       50874.0  ...                         1273269.0   

          Less than 9th grade        9th to 12th grade, no diploma  \
7                        0.14                                 0.14   
10                       0.10                                 0.08   
1                        0.13                                 0.14   
4                        0.10                                 0.08   
9                        0.07                                 0.07   

          High school graduate (includes equivalency)  \
7                                                0.28   
10                                               0.29   
1                                                0.28   
4                                                0.27   
9                                                0.12   

          Some college, no degree        Associate's degree  \
7                            0.17                      0.07   
10                           0.14                      0.08   
1                            0.17                      0.08   
4                            0.14                      0.08   
9                            0.09                      0.04   

          Bachelor's degree        Graduate or professional degree  \
7                      0.13                                   0.07   
10                     0.20                                   0.11   
1                      0.13                                   0.07   
4                      0.21                                   0.12   
9                      0.31                                   0.30   

          Percent high school graduate or higher  \
7                                           0.72   
10                                          0.82   
1                                           0.73   
4                                           0.83   
9                                           0.86   

          Percent bachelor's degree or higher  
7                                        0.20  
10                                       0.31  
1                                        0.21  
4                                        0.34  
9                                        0.61  

[5 rows x 308 columns]

In [46]:
extra_covariates = pd.read_csv('assets/Literally_Everything.csv')
#Get rid of white space in columns
all_covariates = [x.strip() for x in external_features_df.columns] #Get rid of white space
external_features_df.columns = all_covariates
#Keep for reducing columns
external_features_df.head()

Borough    YEAR  Total housing units  Occupied housing units  \
0        NYC  2018.0            3519453.0               3184496.0   
1      Bronx  2018.0             532509.0                507370.0   
2   Brooklyn  2018.0            1053667.0                969317.0   
3  Manhattan  2018.0             886282.0                752258.0   
4     Queens  2018.0             865809.0                788110.0   

   Vacant housing units  1-unit, detached  1-unit, attached   2 units  \
0              334957.0          320628.0          239273.0  460326.0   
1               25139.0           32229.0           22889.0   47809.0   
2               84350.0           54761.0           93895.0  182912.0   
3              134024.0            9221.0            5143.0    9161.0   
4               77699.0          163013.0           71163.0  177611.0   

   3 or 4 units  5 to 9 units  ...  Population 25 years and over  \
0      318507.0      233665.0  ...                     5948745.0   
1       40522.0       20169.0  ...                      934391.0   
2      165442.0      110951.0  ...                     1785323.0   
3       19451.0       48589.0  ...                     1249838.0   
4       87785.0       51055.0  ...                     1646542.0   

   Less than 9th grade  9th to 12th grade, no diploma  \
0                 0.09                           0.08   
1                 0.13                           0.14   
2                 0.09                           0.08   
3                 0.07                           0.05   
4                 0.10                           0.08   

   High school graduate (includes equivalency)  Some college, no degree  \
0                                         0.24                     0.13   
1                                         0.28                     0.17   
2                                         0.26                     0.13   
3                                         0.13                     0.10   
4                                         0.27                     0.14   

   Associate's degree  Bachelor's degree  Graduate or professional degree  \
0                0.06               0.23                             0.16   
1                0.08               0.13                             0.07   
2                0.06               0.23                             0.16   
3                0.04               0.32                             0.29   
4                0.08               0.21                             0.12   

   Percent high school graduate or higher  Percent bachelor's degree or higher  
0                                    0.83                                 0.39  
1                                    0.73                                 0.21  
2                                    0.83                                 0.39  
3                                    0.88                                 0.61  
4                                    0.83                                 0.34  

[5 rows x 308 columns]

In [47]:
boro_features_to_num = {}
for cnt, col_name in enumerate(external_features_df.columns):
    boro_features_to_num[col_name] = str(cnt)
num_to_boro_features = dict((v,k) for k,v in boro_features_to_num.items()) # dict to get feature from the col number
external_features_df = external_features_df.rename(columns=boro_features_to_num)
external_features_df.head()

0       1          2          3         4         5         6  \
0        NYC  2018.0  3519453.0  3184496.0  334957.0  320628.0  239273.0   
1      Bronx  2018.0   532509.0   507370.0   25139.0   32229.0   22889.0   
2   Brooklyn  2018.0  1053667.0   969317.0   84350.0   54761.0   93895.0   
3  Manhattan  2018.0   886282.0   752258.0  134024.0    9221.0    5143.0   
4     Queens  2018.0   865809.0   788110.0   77699.0  163013.0   71163.0   

          7         8         9  ...        298   299   300   301   302   303  \
0  460326.0  318507.0  233665.0  ...  5948745.0  0.09  0.08  0.24  0.13  0.06   
1   47809.0   40522.0   20169.0  ...   934391.0  0.13  0.14  0.28  0.17  0.08   
2  182912.0  165442.0  110951.0  ...  1785323.0  0.09  0.08  0.26  0.13  0.06   
3    9161.0   19451.0   48589.0  ...  1249838.0  0.07  0.05  0.13  0.10  0.04   
4  177611.0   87785.0   51055.0  ...  1646542.0  0.10  0.08  0.27  0.14  0.08   

    304   305   306   307  
0  0.23  0.16  0.83  0.39  
1  0.13  0.07  0.73  0.21  
2  0.23  0.16  0.83  0.39  
3  0.32  0.29  0.88  0.61  
4  0.21  0.12  0.83  0.34  

[5 rows x 308 columns]

In [48]:
# new_york_listings.columns

In [49]:
boro_to_num = {'Bronx':0, 'Brooklyn':1, 'Manhattan':2, 'Queens':3, 'Staten Island':4}
num_to_boro = {0:'Bronx', 1:'Brooklyn', 2:'Manhattan', 3:'Queens', 4:'Staten Island'}
def assign_borough(df, listing_id):
    try:
        return boro_to_num[df['neighbourhood_group_cleansed'].loc[listing_id]]
    except:
        return np.nan

reg1_df['borough'] = reg1_df['listing_id'].apply(lambda x: assign_borough(new_york_listings, x))
reg1_df

#### This cell is supposed to add the external features to the reg df. Perhaps we should run PCA first

listing_id  price  minimum_nights  maximum_nights  month  year  \
0              55982   85.0             3.0            18.0      8  2021   
97             15341  192.0            30.0          1125.0      8  2021   
98             15341  192.0            30.0          1125.0      8  2021   
99             15341  190.0            30.0          1125.0      8  2021   
100            15341  190.0            30.0          1125.0      8  2021   
...              ...    ...             ...             ...    ...   ...   
12142133    49003697  150.0             3.0          1125.0      7  2022   
12142134    49003697  150.0             3.0          1125.0      7  2022   
12142135    49003697  150.0             3.0          1125.0      8  2022   
12142136    49003697  150.0             3.0          1125.0      8  2022   
12142137    49003697  150.0             3.0          1125.0      8  2022   

          host_listings_count host_has_profile_pic host_identity_verified  \
0                         1.0                    t                      t   
97                        1.0                    t                      t   
98                        1.0                    t                      t   
99                        1.0                    t                      t   
100                       1.0                    t                      t   
...                       ...                  ...                    ...   
12142133                  0.0                    t                      t   
12142134                  0.0                    t                      t   
12142135                  0.0                    t                      t   
12142136                  0.0                    t                      t   
12142137                  0.0                    t                      t   

          bedrooms  ...  private_entrance  patio_or_balcony  Pool  \
0              1.0  ...               0.0               0.0   0.0   
97             1.0  ...               0.0               1.0   0.0   
98             1.0  ...               0.0               1.0   0.0   
99             1.0  ...               0.0               1.0   0.0   
100            1.0  ...               0.0               1.0   0.0   
...            ...  ...               ...               ...   ...   
12142133       1.0  ...               0.0               0.0   0.0   
12142134       1.0  ...               0.0               0.0   0.0   
12142135       1.0  ...               0.0               0.0   0.0   
12142136       1.0  ...               0.0               0.0   0.0   
12142137       1.0  ...               0.0               0.0   0.0   

          laundromat_nearby  host_greets_you  free_street_parking  \
0                       0.0              1.0                  1.0   
97                      1.0              0.0                  0.0   
98                      1.0              0.0                  0.0   
99                      1.0              0.0                  0.0   
100                     1.0              0.0                  0.0   
...                     ...              ...                  ...   
12142133                0.0              0.0                  0.0   
12142134                0.0              0.0                  0.0   
12142135                0.0              0.0                  0.0   
12142136                0.0              0.0                  0.0   
12142137                0.0              0.0                  0.0   

          lock_on_bedroom_door  day  weekday  borough  
0                          0.0    4        2        1  
97                         0.0    4        2        2  
98                         0.0    5        3        2  
99                         0.0    6        4        2  
100                        0.0    7        5        2  
...                        ...  ...      ...      ...  
12142133                   0.0   30        5        2  
12142134                   0.0   31        6        2  
12142135  

In [53]:
# reg1_df.to_csv('assets/reg1_df.csv')

In [57]:
X = external_features_df.drop(columns=['0','1']).values
scaler = StandardScaler()
scaler.fit_transform(X)
print(X.shape)

pca = PCA(n_components=2, random_state=33)
new_X = pca.fit_transform(X)
print(new_X.shape)  # new_X is the features after PCA
pca.explained_variance_ratio_

(12, 306)
(12, 2)


array([0.9961756 , 0.00271708])

In [58]:
reduced_features = external_features_df[['0','1']]
# PCs_df = pd.DataFrame(new_X, columns=['PC1','PC2'])
# reduced_features['PC1'] = PCs_df['PC1']
# reduced_features['PC2'] = PCs_df['PC2']
reduced_features['PC1'] = new_X[:,0]
reduced_features['PC2'] = new_X[:,1]
reduced_features['1'] = reduced_features['1'].apply(lambda x: int(x))
reduced_features = reduced_features.set_index(['0','1'])
reduced_features

<ipython-input-58-1796c5538fa8>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  reduced_features['PC1'] = new_X[:,0]
<ipython-input-58-1796c5538fa8>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  reduced_features['PC2'] = new_X[:,1]
<ipython-input-58-1796c5538fa8>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guid

PC1            PC2
0             1                                
NYC           2018  1.540168e+07   91648.587797
Bronx         2018 -4.040512e+06  -56007.106796
Brooklyn      2018 -7.461914e+05  -41045.445045
Manhattan     2018 -2.949708e+06  799399.470721
Queens        2018 -1.598534e+06 -379898.275268
Staten Island 2018 -6.555179e+06 -187905.798952
NYC           2017  1.589024e+07  -95568.277532
Bronx         2017 -3.960579e+06 -111670.351392
Brooklyn      2017 -6.174912e+05 -125683.228725
Manhattan     2017 -2.848440e+06  773519.117500
Queens        2017 -1.429840e+06 -451931.232557
Staten Island 2017 -6.545452e+06 -214857.459750

In [75]:
def assign_PCs(df, borough, year, pc):
    try:
#         print(num_to_boro[int(borough)])
#         print(int(year)-3)
        return df.loc[num_to_boro[int(borough)],int(year)-4][pc]
    except:
        return np.nan

reg2_df = reg1_df # [:10]
for pc in ['PC1','PC2']:
    reg2_df[pc] = reg2_df.apply(lambda x: assign_PCs(reduced_features, x['borough'], x['year'], pc), axis=1)
reg2_df

listing_id  price  minimum_nights  maximum_nights  month  year  \
0              55982   85.0             3.0            18.0      8  2021   
97             15341  192.0            30.0          1125.0      8  2021   
98             15341  192.0            30.0          1125.0      8  2021   
99             15341  190.0            30.0          1125.0      8  2021   
100            15341  190.0            30.0          1125.0      8  2021   
...              ...    ...             ...             ...    ...   ...   
12142133    49003697  150.0             3.0          1125.0      7  2022   
12142134    49003697  150.0             3.0          1125.0      7  2022   
12142135    49003697  150.0             3.0          1125.0      8  2022   
12142136    49003697  150.0             3.0          1125.0      8  2022   
12142137    49003697  150.0             3.0          1125.0      8  2022   

          host_listings_count host_has_profile_pic host_identity_verified  \
0                         1.0                    t                      t   
97                        1.0                    t                      t   
98                        1.0                    t                      t   
99                        1.0                    t                      t   
100                       1.0                    t                      t   
...                       ...                  ...                    ...   
12142133                  0.0                    t                      t   
12142134                  0.0                    t                      t   
12142135                  0.0                    t                      t   
12142136                  0.0                    t                      t   
12142137                  0.0                    t                      t   

          bedrooms  ...  Pool  laundromat_nearby  host_greets_you  \
0              1.0  ...   0.0                0.0              1.0   
97             1.0  ...   0.0                1.0              0.0   
98             1.0  ...   0.0                1.0              0.0   
99             1.0  ...   0.0                1.0              0.0   
100            1.0  ...   0.0                1.0              0.0   
...            ...  ...   ...                ...              ...   
12142133       1.0  ...   0.0                0.0              0.0   
12142134       1.0  ...   0.0                0.0              0.0   
12142135       1.0  ...   0.0                0.0              0.0   
12142136       1.0  ...   0.0                0.0              0.0   
12142137       1.0  ...   0.0                0.0              0.0   

          free_street_parking  lock_on_bedroom_door  day  weekday  borough  \
0                         1.0                   0.0    4        2        1   
97                        0.0                   0.0    4        2        2   
98                        0.0                   0.0    5        3        2   
99                        0.0                   0.0    6        4        2   
100                       0.0                   0.0    7        5        2   
...                       ...                   ...  ...      ...      ...   
12142133                  0.0                   0.0   30        5        2   
12142134                  0.0                   0.0   31        6        2   
12142135                  0.0                   0.0    1        0        2   
12142136                  0.0                   0.0    2        1        2   
12142137                  0.0                   0.0    3        2        2   

                   PC1            PC2  
0        -6.174912e+05 -125683.228725  
97       -2.848440e+06  773519.117500  
98       -2.848440e+06  773519.117500  
99       -2.848440e+06  773519.117500  
100      -2.848440e+06  773519.117500  
...                ...            ...  
12142133 -2.949708e+06  799399.470721  
12142134 -2.949708e+06  799399.470721  
12142135 -2.949708e+06  799399.470721  
121421

In [77]:
# reg2_df[np.isnan(reg2_df['PC1'])]

In [78]:
# scaler = StandardScaler()
# X = reg2_df.values
# X

In [79]:
formula = "price ~ minimum_nights + maximum_nights + month + year + day + weekday + host_listings_count + \
           host_has_profile_pic + host_identity_verified + bedrooms + beds + availability_60 + \
           number_of_reviews_l30d + review_scores_rating + TV + Dryer + Shampoo + Gym + Pool + private_entrance + \
           patio_or_balcony + laundromat_nearby + host_greets_you + free_street_parking + lock_on_bedroom_door + \
           PC1 + PC2"

reg2 = smf.ols(formula=formula, data=reg2_df).fit()
reg2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.105
Model:                            OLS   Adj. R-squared:                  0.105
Method:                 Least Squares   F-statistic:                 3.184e+04
Date:                Mon, 23 Aug 2021   Prob (F-statistic):               0.00
Time:                        20:27:28   Log-Likelihood:            -5.1917e+07
No. Observations:             7328553   AIC:                         1.038e+08
Df Residuals:                 7328525   BIC:                         1.038e+08
Df Model:                          27                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                   -1637.4695    836.251     -1.958      0.050   -3276.491       1.552
host_has_profile_pic[T.t]     -15.6406      2.259     -6.925      0.000     -20.067     -11.214
host_identity_verified[T.t]     2.3400      0.274      8.530      0.000       1.802       2.878
minimum_nights                 -0.0754      0.003    -22.381      0.000      -0.082      -0.069
maximum_nights               1.002e-07   1.52e-09     65.822      0.000    9.72e-08    1.03e-07
month                           0.8564      0.059     14.533      0.000       0.741       0.972
year                            0.8029      0.413      1.942      0.052      -0.008       1.613
day                             0.0180      0.012      1.482      0.138      -0.006       0.042
weekday                         0.4451      0.053      8.338      0.000       0.340       0.550
host_listings_count             0.0251      0.002     14.355      0.000       0.022       0.029
bedrooms                       80.0865      0.229    350.197      0.000      79.638      80.535
beds                           14.8754      0.135    110.534      0.000      14.612      15.139
availability_60                 0.9961      0.005    214.653      0.000       0.987       1.005
number_of_reviews_l30d         -0.6582      0.078     -8.419      0.000      -0.811      -0.505
review_scores_rating            0.0199      0.012      1.655      0.098      -0.004       0.044
TV                             18.2427      0.243     75.047      0.000      17.766      18.719
Dryer                          12.1954      0.234     52.153      0.000      11.737      12.654
Shampoo                        26.2995      0.239    109.917      0.000      25.831      26.768
Gym                            51.4310      0.429    119.967      0.000      50.591      52.271
Pool                           98.2118      1.086     90.408      0.000      96.083     100.341
private_entrance               17.7254      0.273     64.972      0.000      17.191      18.260
patio_or_balcony               36.7718      0.348    105.680      0.000      36.090      37.454
laundromat_nearby              -0.6732      0.495     -1.361      0.174      -1.643       0.296
host_greets_you                 0.5610      0.301      1.866      0.062      -0.028       1.150
free_street_parking           -37.1333      0.244   -152.477      0.000     -37.611     -36.656
lock_on_bedroom_door          -37.6262      0.259   -145.242      0.000     -38.134     -37.118
PC1                          7.435e-06   1.24e-07     59.918      0.000    7.19e-06    7.68e-06
PC2                          8.619e-05   3.21e-07    268.106      0.000    8.56e-05    8.68e-05
==============================================================================
Omnibus:                 177

In [82]:
reg2.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [93]:
compare_amenities(new_york_listings, 0.005)

alt.Chart(...)